### ISSUE TITLE
Upgrade from 1.23 to 1.24 incomplete

### ISSUE THREAD

**Issue by mburket** [istio/istio#53875]
### Is this the right place to submit this?

- [X] This is not a security vulnerability or a crashing bug
- [X] This is not a question about how to use Istio

### Bug Description

We are running version 1.23.2 of Istio and we have two ingress-gateways installed. We built our operator manifest for installation. When we tried to upgrade to 1.24.0, the installation never completed and it got stuck at installing the second gateway which has deployment and service overlay. So we ended up with one gateway updated to 1.24 and the other is still on 1.23. No debug logs showed me why it was stuck. Please advise. 

### Version

```Text
❯ istioctl version
client version: 1.24.0
istiod version: 1.23.2
istiod version: 1.24.0
data plane version: 1.23.2 (2 proxies), 1.24.0 (26 proxies)
> ❯ kubectl version
Client Version: v1.31.1
Kustomize Version: v5.4.2
Server Version: v1.31.2-eks-7f9249a
```

**Comment by ilrudie** [comment]
Could you provide some details about how you were trying to perform the upgrade? Did you attempt an in-place upgrade or revisioned? Is it possible to provide the IOP resources you used? Did one of your gateways not start successfully or did it come up but the installation didn't seem to correctly detect it?

**Comment by mburket** [comment]
@ilrudie We used Canary upgrade and we installed 1.24 alongside with 1.23. Both ingress-gateways camp up, but the external-ingress-gateway didn't upgrade. I checked this GW's deployment and service, and they are still on version 1.23. Below is the operator file we use for installation. The installation stuck at ingress-gateway installed:

This will install the Istio 1.24.0 profile "default" into the cluster. Proceed? (y/N) y
✔ Istio core installed ⛵️
✔ Istiod installed 🧠
✔ CNI installed 🪢
✔ Ingress gateways installed 🛬
<Never finish here...>
^C

---
apiVersion: install.istio.io/v1alpha1
kind: IstioOperator
metadata:
  namespace: istio-system
  name: istiocontrolplane
spec:
  revision: 1-24-0

  profile: default

  components:
    cni:
      enabled: true
      k8s:
        affinity:
          nodeAffinity:
            requiredDuringSchedulingIgnoredDuringExecution:
              nodeSelectorTerms:
              - matchExpressions:
                - key: eks.amazonaws.com/compute-type
                  operator: NotIn
                  values:
                  - fargate

    pilot:
      k8s:
        serviceAnnotations:
          ad.datadoghq.com/discovery.check_names: '["istio"]'
          ad.datadoghq.com/discovery.init_configs: '[{}]'
          ad.datadoghq.com/discovery.instances: |
              [
                {
                  "cluster_check": true,
                  "istiod_endpoint": "http://%%host%%:15014/metrics",
                  "use_openmetrics": "true"
                  "exclude_labels": [
                    "source_version",
                    "destination_version",
                    "source_canonical_revision",
                    "destination_canonical_revision",
                    "source_principal",
                    "destination_principal",
                    "source_cluster",
                    "destination_cluster",
                    "source_canonical_service",
                    "destination_canonical_service",
                    "source_workload_namespace",
                    "destination_workload_namespace",
                    "request_protocol",
                    "connection_security_policy"
                  ]
                }
              ]

    ingressGateways:
      - name: istio-ingressgateway
        enabled: true
        label:
          istio: istio-ingressgateway
        k8s:
          overlays:
            - kind: Service
              name: istio-ingressgateway
              patches:
                # Set a expected nodePort to enable ALB ingress
                - path: spec.ports.[name:status-port].nodePort
                  value: 30021
                - path: spec.ports.[name:http2].nodePort
                  value: 30080
                - path: spec.ports.[name:https].nodePort
                  value: 30443
      - name: istio-external-ingressgateway
        enabled: true
        label:
          istio: istio-external-ingressgateway
        k8s:
          overlays:
            - kind: Deployment
              name: istio-external-ingressgateway
              patches:
                - path: metadata.labels.app
                  value: istio-external-ingressgateway
                - path: metadata.labels.istio
                  value: external-ingressgateway
                - path: spec.selector.matchLabels.app
                  value: istio-external-ingressgateway
                - path: spec.selector.matchLabels.istio
                  value: external-ingressgateway
                - path: spec.template.metadata.labels.app
                  value: istio-external-ingressgateway
                - path: spec.template.metadata.labels.istio
                  value: external-ingressgateway
            - kind: Service
              name: istio-external-ingressgateway
              patches:
                - path: metadata.labels.app
                  value: istio-external-ingressgateway
                - path: metadata.labels.istio
                  value: external-ingressgateway
                - path: spec.selector.app
                  value: istio-external-ingressgateway
                - path: spec.selector.istio
                  value: external-ingressgateway
                # Set a expected nodePort to enable ALB ingress
                - path: spec.ports.[name:status-port].nodePort
                  value: 31021
                - path: spec.ports.[name:http2].nodePort
                  value: 31080
                - path: spec.ports.[name:https].nodePort
                  value: 31443
  values:
    global:
      meshID: dev
      multiCluster:
        clusterName: dev-kubernetes-west
      network: dev

    gateways:
      istio-ingressgateway:
        autoscaleEnabled: true
        type: NodePort
        name: istio-ingressgateway
        env: {}
        secretVolumes:
          - name: ingressgateway-certs
            secretName: istio-ingressgateway-certs
            mountPath: /etc/istio/ingressgateway-certs
          - name: ingressgateway-ca-certs
            secretName: istio-ingressgateway-ca-certs
            mountPath: /etc/istio/ingressgateway-ca-certs
      istio-ingressgateway:
        autoscaleEnabled: true
        type: NodePort
        name: istio-external-ingressgateway
        env: {}
        secretVolumes:
          - name: ingressgateway-certs
            secretName: istio-ingressgateway-certs
            mountPath: /etc/istio/ingressgateway-certs
          - name: ingressgateway-ca-certs
            secretName: istio-ingressgateway-ca-certs
            mountPath: /etc/istio/ingressgateway-ca-certs

**Comment by ilrudie** [comment]
Thanks, can repro in KinD with a simple `$ istioctl install -f sample-iop.yaml -y` where sample-iop.yaml is the file provided above. No need to even update, regular install hangs.

**Comment by ilrudie** [comment]
99% sure I see the issue, building a new istioctl with the fix locally and then will file a PR if it works

**Comment by mburket** [comment]
@ilrudie Correct, i observed the same hanging behavior for a fresh installation in a different cluster. Btw, we have been using the same operator file for previous versions.

**Comment by ilrudie** [comment]
gotcha, yeah I'm pretty sure it's a bug in istioctl but it looks like it is already fixed on the tip of master. just going to try to ID the PR which fixed it and make sure it has been cherry-picked

**Comment by ilrudie** [comment]
actually I can repro on tip of master too, will dig into it more tomorrow and try to figure it out

**Comment by ilrudie** [comment]
It's a concurrency issue in the install logic. I have a code fix I will PR and will try to get it into 1.24.1 which is being planned now.

**Comment by ilrudie** [comment]
#53930 merges this fix into our 1.24 branch so the fix will make it into 1.24.1

**Comment by ilrudie** [comment]
[Istio 1.24.1](https://istio.io/latest/news/releases/1.24.x/announcing-1.24.1/) has been released!
